packages/nimble-components/src/table/index.ts

```diff
@@ -336,35 +373,103 @@ export class Table<
+            if (this.collapsedState[collapsedRowId]) {
+                expandedState[collapsedRowId] = false;
+            }
+        }
         this.updateTableOptions({
-            data
+            state: {
+                expanded: expandedState
+            }
         });
     }
 
+    private getInflatedExpandedState(): TanStackExpandedState {
+        if (this.table.options.state.expanded === true) {
+            const expandedState = {} as { [key: string]: boolean };
+            for (const row of this.table.getRowModel().rows) {
+                expandedState[row.id] = true;
+            }
+            return expandedState;
+        }
+
+        return this.table.options.state.expanded!;
+    }
+
     private refreshRows(): void {
         const rows = this.table.getRowModel().rows;
-        this.tableData = rows.map(row => {
+        this.tableData = rows.map((row, i) => {
+            const groupColumn = this.getGroupRowColumn(i);
             const rowState: TableRowState<TData> = {
                 record: row.original,
-                id: row.id
+                id: row.id,
+                isGrouped: row.getIsGrouped(),
+                isExpanded: row.getIsExpanded(),
+                groupRowValue: groupColumn
+                    ? row?.getLeafRows()[0]?.getValue(row.groupingColumnId!)
+                    : undefined,
+                nestingLevel: row.depth,
+                leafItemCount: row
+                    .getLeafRows()
+                    .filter(leafRow => leafRow.getLeafRows().length === 0)
+                    .length,
+                groupColumn
             };
             return rowState;
         });
         this.virtualizer.dataChanged();
     }
 
+    private getGroupRowColumn(rowIndex: number): TableColumn | undefined {
+        const groupedId = this.table.getRowModel()!.rows[rowIndex]?.groupingColumnId;
+        if (groupedId !== undefined) {
+            return this.columns.find(c => c.internalUniqueId === groupedId);
+        }
+
+        return undefined;
+    }
+
     private updateTableOptions(
-        updatedOptions: Partial<TanStackTableOptionsResolved<TData>>
+        updatedOptions: Partial<TanStackTableOptionsResolved<TData>>,
+        skipRowRefresh?: boolean
     ): void {
         this.options = {
             ...this.options,
             ...updatedOptions,
             state: { ...this.options.state, ...updatedOptions.state }
         };
         this.table.setOptions(this.options);
-        this.refreshRows();
+        if (!skipRowRefresh) {
+            this.refreshRows();
+        }
     }
 
+    private readonly handleExpandedChange = (updater: unknown): void => {
+        let expandedState = this.table.options.state.expanded;
+        expandedState = updater instanceof Function
+            ? (updater(expandedState) as TanStackExpandedState)
+            : (expandedState = updater as TanStackExpandedState);
```

Comment:
Why are you assigning to `expandedState` here? Isn't the assignment handled two lines above?